### ISSUE TITLE
re-fix treatment of bad fits when ranking minima

### ISSUE THREAD

**Issue by sbailey** [desihub/redrock#225]
PR #211 (new GPU code) had some merge conflicts with the changes from PR #218 (avoid bad fits when ranking minima).  It appears that this wasn't completely resolved correctly, but we didn't catch it before merging #211 and the bad fits have reappeared.

A file with two example targets from sv3 dark healpix 10152 is in /global/cfs/cdirs/desi/users/sjbailey/dev/redrock/gpu/zrank/coadd-badminfit.fits.  I ran with the previous tag (0.16.0) and the latest main:
```
module swap redrock/0.16.0
time srun -n 4 -c 2 --gpu-bind=map_gpu:3,2,1,0 --cpu-bind=cores \
    rrdesi_mpi --gpu --max-gpuprocs 4 -i coadd-badminfit.fits -o redrock-old.fits -d rrdetails-old.h5

module swap redrock/main
time srun -n 4 -c 2 --gpu-bind=map_gpu:3,2,1,0 --cpu-bind=cores \
    rrdesi_mpi --gpu --max-gpuprocs 4 -i coadd-badminfit.fits -o redrock-gpu.fits -d rrdetails-gpu.h5
```

Comparing the top 9 fits from the rrdetails files for these two runs:
```
import h5py
import numpy as np
from astropy.table import Table

h1 = h5py.File('rrdetails-old.h5')
h2 = h5py.File('rrdetails-gpu.h5')

targetid = 39633308400748798
zfit1 = Table(np.array(h1["zfit"][str(targetid)]["zfit"]))
zfit2 = Table(np.array(h2["zfit"][str(targetid)]["zfit"]))
print(zfit1["znum", "z", "spectype", "chi2", "zwarn"])
print(zfit2["znum", "z", "spectype", "chi2", "zwarn"])
```

With tag 0.16.0 (previous main):
```
znum            z            spectype        chi2        zwarn
---- ----------------------- -------- ------------------ -----
   0      0.7465644325737828   GALAXY  8031.702453918755     4
   1     0.02090719822607346   GALAXY  8035.408617666035     4
   2   0.0012759003749705934     STAR  8116.358099282074     4
   3      1.0706405066878273      QSO  8125.152400165796     4
   4  -0.0005147214510569681     STAR  8129.311497150431     4
   5      1.0334574452161545      QSO 8133.4641371942125     4
   6       2.086404894124313      QSO  8134.274736442487     4
   7 -0.00048533154378753955     STAR  8158.796972346495     0
   8      1.6978828086448678   GALAXY  8029.752568023694  1060
```

with current main (just merged from PR #211):
```
znum           z            spectype        chi2        zwarn
---- ---------------------- -------- ------------------ -----
   0     1.6978828086448678   GALAXY 8029.7525680238305  1060
   1      0.746564432573787   GALAXY  8031.702453941107     4
   2   0.020907198226076246   GALAXY  8035.408617666151     4
   3   0.001275900374917386     STAR  8116.358099282008     4
   4     1.0706405066878273      QSO  8125.152400163468     4
   5 -0.0005147214510649985     STAR   8129.31149715046     4
   6     1.0334574452161573      QSO  8133.464137193747     4
   7     2.0864048941242976      QSO  8134.274736442952     4
   8 -0.0004853315438139842     STAR  8158.796972346274     0
```

Note that the GALAXY Z=1.69788 fit has zwarn bit 10 (BAD_MINFIT) set.  Even though it has the lowest chi2, PR #218 (tag 0.16.0) correctly sent this to the back of the list because that minfit (chi2) can't be trusted.  Current main ranks this as the best fit.

targetid=616094057126429663 is the other case in this file, with the same symptom.

Please
1. write a unit test that would have caught this bug (i.e. to prevent us from breaking it again in a future update)
2. fix the bug

Thanks!

**Comment by craigwarner-ufastro** [comment]
Simple bug - np.lexsort((a,b)) results in sorting by column b first then column a.  This is counterintuitive and opposite to behavior of astropy.Table.sort((a,b)).  This has been brought in line with expected sort order by adding a boolean variable to sort_dict_by_cols, which will sort by the first column first - this is set to True by default.

Also added unit test `test_sort_zfit_dict` that is analogous to `test_sort_zfit` and also added `test_sort_dict_by_cols`.

**Comment by sbailey** [comment]
fixed by PR #226
